rust/src/http2/http2.rs

```diff
@@ -0,0 +1,1158 @@
+            if self.request_frame_size >= ilen {
+                self.request_frame_size -= ilen;
+                return AppLayerResult::ok();
+            } else {
+                let start = self.request_frame_size as usize;
+                input = &input[start..];
+                self.request_frame_size = 0;
+            }
+        }
+
+        //then parse all we can
+        while input.len() > 0 {
+            match parser::http2_parse_frame_header(input) {
+                Ok((rem, head)) => {
+                    let hl = head.length as usize;
+
+                    //we check for completeness first
+                    if rem.len() < hl {
+                        //but limit ourselves so as not to exhaust memory
+                        if hl < HTTP2_MAX_HANDLED_FRAME_SIZE {
+                            return AppLayerResult::incomplete(
+                                (il - input.len()) as u32,
+                                (HTTP2_FRAME_HEADER_LEN + hl) as u32,
+                            );
+                        } else if rem.len() < HTTP2_MIN_HANDLED_FRAME_SIZE {
+                            return AppLayerResult::incomplete(
+                                (il - input.len()) as u32,
+                                (HTTP2_FRAME_HEADER_LEN + HTTP2_MIN_HANDLED_FRAME_SIZE) as u32,
+                            );
+                        } else {
+                            self.set_event(HTTP2Event::LongFrameData);
+                            self.request_frame_size = head.length - (rem.len() as u32);
+                        }
+                    }
+
+                    //get a safe length for the buffer
+                    let (hlsafe, complete) = if rem.len() < hl {
+                        (rem.len(), false)
+                    } else {
+                        (hl, true)
+                    };
+
+                    if head.length == 0 && head.ftype == parser::HTTP2FrameType::SETTINGS as u8 {
+                        input = &rem[hlsafe..];
+                        continue;
+                    }
+                    let txdata = self.parse_frame_data(
+                        head.ftype,
+                        &rem[..hlsafe],
+                        complete,
+                        head.flags,
+                        STREAM_TOSERVER,
+                    );
+
+                    let tx = self.find_or_create_tx(&head, &txdata, STREAM_TOSERVER);
+                    tx.handle_frame(&head, &txdata, STREAM_TOSERVER);
+                    let over = head.flags & parser::HTTP2_FLAG_HEADER_EOS != 0;
+                    let txid = tx.tx_id;
+                    let ftype = head.ftype;
+                    tx.frames_ts.push(HTTP2Frame {
+                        header: head,
+                        data: txdata,
+                    });
+                    if ftype == parser::HTTP2FrameType::DATA as u8 {
+                        self.stream_data(STREAM_TOSERVER, &rem[..hlsafe], over, txid);
```

Comment:
does this assume that the data is always a file? Can't we have different content types, like chunked encoding or multipart? If so, that needs to be handled before passing data to the file API

Reply:
> does this assume that the data is always a file?

Yes it does

> Can't we have different content types, like chunked encoding or multipart?

I do not know, not in the first pcap.
This needs more testing